src/crypto/tls/handshake_server_tls13.go

```diff
@@ -169,12 +170,39 @@ func (hs *serverHandshakeStateTLS13) processClientHello() error {
 	hs.hello.cipherSuite = hs.suite.id
 	hs.transcript = hs.suite.hash.New()
 
+	// Resolve the server's preference for the ECDHE group.
+	supportedCurves := c.config.curvePreferences()
+	if testingTriggerHRR {
+		// A HelloRetryRequest (HRR) is sent if the client does not offer a key
+		// share for a curve supported by the server. To trigger this condition
+		// intentionally, we compute the set of ECDHE groups supported by both
+		// the client and server but for which the client did not offer a key
+		// share.
+		supportedCurves = make([]CurveID, 0)
+		for _, serverGroup := range c.config.curvePreferences() {
+			for _, clientGroup := range hs.clientHello.supportedCurves {
+				if serverGroup == clientGroup {
+					for _, ks := range hs.clientHello.keyShares {
+						if serverGroup != ks.group {
+							supportedCurves = append(supportedCurves, serverGroup)
+						}
+					}
+				}
+			}
+		}
+		if len(supportedCurves) == 0 {
+			// This occurs if the client offered a key share for each mutually
+			// supported group.
+			panic("failed to trigger HelloRetryRequest")
```

Comment:
This is fenced by `testingTriggerHRR`, which can only be set within the package and is only used for testing purposes.

The `panic()` is appropriate because the server doesn't know how to proceed.

Reply:
Ah, okay, as long as that's only enabled during tests.